### PR TITLE
[WM-2500][WM-2502] Fetch Github token from ECM for importing and running private workflows

### PR DESCRIPTION
**Pull request by salonishah11**
This PR adds support for fetching Github token for user from ECM to support private workflows.

Testing
I have tested these changes in Workflows and Cromwell Runner apps in dev by manually editing the config as well as updating the deployment with Cromwell image containing the changes. Screenshot of a successful run of private workflow.

image

Closes 
- https://broadworkbench.atlassian.net/browse/WM-2500
- https://broadworkbench.atlassian.net/browse/WM-2502

### CODE REVIEW

File: services/src/main/scala/cromwell/services/auth/ecm/EcmService.scala [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Success, Try}
+import spray.json._
+
+class EcmService(baseEcmUrl: String) {
+  private val getGithubAccessTokenApiPath = "api/oauth/v1/github/access-token"
+
+  /*
+     ECM doesn't have a standard error response format. Some of the responses contains HTML tags in it. This helper
```
Comment (cjllanwarne) [inline services/src/main/scala/cromwell/services/auth/ecm/EcmService.scala:17]: I think technically ECM always returns json format, but it's possible for things in the network layer between you and ECM (like the apache proxies, etc) to return html pages

File: services/src/main/scala/cromwell/services/auth/ecm/EcmService.scala [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        Try(responseBodyAsStr.parseJson) match {
+          case Success(JsObject(fields)) =>
+            fields.get("message").map(_.toString().replaceAll("\"", "")).getOrElse(responseBodyAsStr)
+          case _ => responseBodyAsStr
+        }
+      case _ => responseBodyAsStr
+    }
+
+  def getGithubAccessToken(
+    userToken: String
```
Comment (cjllanwarne) [inline services/src/main/scala/cromwell/services/auth/ecm/EcmService.scala:38]: Minor: to avoid confusion and mixups, it might be nice to have a couple of case classes like
```scala
case class GithubToken(value: String)
case class TerraToken(value: String)
```

for passing these around. Otherwise I can easily foresee it being kind of hard to follow around all the various "userToken"s that are just string typed

File: services/src/main/scala/cromwell/services/auth/impl/GithubAuthVendingActor.scala [hunk 3]
```diff
@@ -1,26 +1,51 @@
...
     extends Actor
     with LazyLogging {
 
-  lazy val enabled = serviceConfig.getBoolean("enabled")
+  implicit val system: ActorSystem = context.system
+  implicit val ec: ExecutionContext = context.dispatcher
+
+  lazy val enabled: Boolean = serviceConfig.getBoolean("enabled")
+
+  private lazy val ecmConfigOpt: Option[EcmConfig] = EcmConfig.apply(serviceConfig)
```
Comment (cjllanwarne) [inline services/src/main/scala/cromwell/services/auth/impl/GithubAuthVendingActor.scala:28]: FYI this should work but `apply` is a pseudo-reserved word for scala because it gets auto-called when you run `var = EcmConfig(serviceConfig)`. So the convention would be for an `apply` to return an instance, not an option of an instance.

In this case, I would just rename to something like `fromConfig`

Comment (cjllanwarne) [inline services/src/main/scala/cromwell/services/auth/impl/GithubAuthVendingActor.scala:28]: Because it's in `reference.conf` there will always be _some_ kind of ecm config in the config (you can't remove values by adding a config on top, only redefined them).

I think what you really want is for the baseUrl to be optional, and for _that_ to be what we're switching on.

File: services/src/main/scala/cromwell/services/auth/impl/GithubAuthVendingActor.scala [hunk 4]
```diff
@@ -1,26 +1,51 @@
...
+          respondTo ! GithubAuthVendingFailure(
+            "Invalid configuration for service 'GithubAuthVending': missing 'ecm.base-url' value."
+          )
+      }
     // This service currently doesn't do any work on shutdown but the service registry pattern requires it
     // (see https://github.com/broadinstitute/cromwell/issues/2575)
     case ShutdownCommand => context stop self
-    case _ =>
-      sender() ! NoGithubAuthResponse
+    case _ => sender() ! NoGithubAuthResponse
```
Comment (cjllanwarne) [inline services/src/main/scala/cromwell/services/auth/impl/GithubAuthVendingActor.scala:48]: I think I would prefer explicitly distinguishing between expected "do nothing" cases and unexpected "oops" cases, eg
```
case GithubAuthRequest(userToken) => sender() ! NoGithubAuthResponse
case other => // Log programmer error and then return either some kind of error response or a noauth response
```

File: services/src/main/scala/cromwell/services/auth/ecm/EcmService.scala [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+  )(implicit actorSystem: ActorSystem, ec: ExecutionContext): Future[String] = {
+
+    def responseEntityToFutureStr(responseEntity: ResponseEntity): Future[String] =
+      responseEntity.dataBytes.runFold(ByteString(""))(_ ++ _).map(_.utf8String)
```
Comment (cjllanwarne) [inline services/src/main/scala/cromwell/services/auth/ecm/EcmService.scala:42]: I know akka is picky and can throw exceptions if we don't read the bytes in time (within 1 second?). How comfortable are we that this future will evaluate in time?

File: services/src/test/scala/cromwell/services/auth/ecm/EcmServiceSpec.scala [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+     "Response error - Bad Gateway"
+    )
+  )
+
+  behavior of "extractErrorMessage in EcmService"
+
+  forAll(testCases) { (testName, statusCode, responseBodyAsStr, expectedErrorMsg) =>
+    it should testName in {
+      assert(ecmService.extractErrorMessage(statusCode, responseBodyAsStr) == expectedErrorMsg)
+    }
```
Comment (cjllanwarne) [inline services/src/test/scala/cromwell/services/auth/ecm/EcmServiceSpec.scala:60]: Woohoo for table based tests!

File: services/src/test/scala/cromwell/services/auth/impl/GithubAuthVendingActorSpec.scala [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+  final private val githubAuthEnabledServiceConfig =
+    ConfigFactory.parseString(s"""
+                                 |enabled = true
+                                 |auth.azure = true
+                                 |ecm.base-url = "https://mock-ecm-url.org"
+        """.stripMargin)
+
+  behavior of "GithubAuthVendingActor"
+
+  it should "return NoGithubAuthResponse if GithubAuthVending is disabled" in {
```
Comment (cjllanwarne) [inline services/src/test/scala/cromwell/services/auth/impl/GithubAuthVendingActorSpec.scala:32]: These also look like a series of very similar tests. Could these also be table based?

File: services/src/main/scala/cromwell/services/auth/impl/GithubAuthVendingActor.scala (hunk 3 above)
Comment (salonishah11) [reply services/src/main/scala/cromwell/services/auth/impl/GithubAuthVendingActor.scala:28]: That's a good point that you can't remove values and only redefine them in additional configs. Then in this case it means that ECM base url will always be present so I think we might not need for either base url or the EcmConfig to be Optional. What do you think about simply removing the Optional and assuming ecmConfig will always be defined?

File: services/src/main/scala/cromwell/services/auth/ecm/EcmService.scala (hunk 5 above)
Comment (salonishah11) [reply services/src/main/scala/cromwell/services/auth/ecm/EcmService.scala:42]: I copied the pattern we have in Cromwell (see [WorkflowCallbackActor.scala](https://github.com/broadinstitute/cromwell/blob/027de62cf8e37f62330d76dba67ed07cce81a1ba/engine/src/main/scala/cromwell/engine/workflow/lifecycle/finalization/WorkflowCallbackActor.scala#L210) and [TesAsyncBackendJobExecutionActor.scala](https://github.com/broadinstitute/cromwell/blob/027de62cf8e37f62330d76dba67ed07cce81a1ba/supportedBackends/tes/src/main/scala/cromwell/backend/impl/tes/TesAsyncBackendJobExecutionActor.scala#L490)) and didn't realize that 1 second was the default timeout. Is there an implicit timeout being imported in these 2 references that I also need to update here? 🤔

File: services/src/test/scala/cromwell/services/auth/impl/GithubAuthVendingActorSpec.scala (hunk 7 above)
Comment (salonishah11) [reply services/src/test/scala/cromwell/services/auth/impl/GithubAuthVendingActorSpec.scala:32]: I will take a stab and try and make them table based tests.

File: services/src/main/scala/cromwell/services/auth/impl/GithubAuthVendingActor.scala (hunk 4 above)
Comment (salonishah11) [reply services/src/main/scala/cromwell/services/auth/impl/GithubAuthVendingActor.scala]: I like the idea of logging programmer error in the second case. I have updated the code.

File: services/src/main/scala/cromwell/services/auth/ecm/EcmService.scala (hunk 1 above)
Comment (salonishah11) [reply services/src/main/scala/cromwell/services/auth/ecm/EcmService.scala]: That's true. I have updated the comment. I also looked at the error handling in ECM and it seems it should return a standard JSON response in format of ErrorReport in all other status codes. Hence I have updated the helper method to
- only return custom error message in case its 401
- try and parse response body as JSON and extract the value associated with `message` key if found
- return the response body in case parsing failed or key `message` was not found

File: services/src/test/scala/cromwell/services/auth/impl/GithubAuthVendingActorSpec.scala (hunk 7 above)
Comment (salonishah11) [reply services/src/test/scala/cromwell/services/auth/impl/GithubAuthVendingActorSpec.scala]: The tests have been updated to be table based.

File: services/src/main/scala/cromwell/services/auth/impl/GithubAuthVendingActor.scala [hunk 8]
```diff
@@ -25,27 +27,32 @@ class GithubAuthVendingActor(serviceConfig: Config, globalConfig: Config, servic
...
 
   override def receive: Receive = {
-    case GithubAuthRequest(userToken) if enabled =>
+    case GithubAuthRequest(terraToken) if enabled =>
       val respondTo = sender()
       ecmServiceOpt match {
         case Some(ecmService) =>
-          ecmService.getGithubAccessToken(userToken) onComplete {
-            case Success(githubToken) => respondTo ! GithubAuthTokenResponse(githubToken)
+          ecmService.getGithubAccessToken(terraToken.value) onComplete {
```
Comment (cjllanwarne) [inline services/src/main/scala/cromwell/services/auth/impl/GithubAuthVendingActor.scala:38]: Should the `ecmService.getGithubAccessToken` be upgraded to take a TerraToken too? ie only unpack it at the very last moment?

File: services/src/main/scala/cromwell/services/auth/impl/GithubAuthVendingActor.scala [hunk 9]
```diff
@@ -25,27 +27,32 @@ class GithubAuthVendingActor(serviceConfig: Config, globalConfig: Config, servic
...
+            case Success(token) => respondTo ! GithubAuthTokenResponse(GithubToken(token))
```
Comment (cjllanwarne) [inline services/src/main/scala/cromwell/services/auth/impl/GithubAuthVendingActor.scala:39]: And return a GithubToken?

File: services/src/test/scala/cromwell/services/auth/impl/GithubAuthVendingActorSpec.scala [hunk 10]
```diff
@@ -8,94 +8,81 @@ import cromwell.services.auth.GithubAuthVending.{
...
-    val serviceRegistryActor = TestProbe()
-    val actor = system.actorOf(
-      Props(
-        new TestGithubAuthVendingActor(githubAuthEnabledServiceConfig,
-                                       ConfigFactory.parseString(""),
-                                       serviceRegistryActor.ref
-        )
-      )
+  private val testCases = Table(
+    ("test name", "service config", "use TestEcmService class", "terra token", "response"),
```
Comment (cjllanwarne) [inline services/src/test/scala/cromwell/services/auth/impl/GithubAuthVendingActorSpec.scala:46]: Can we add a case for "valid config / no token for user in ECM"?

File: services/src/test/scala/cromwell/services/auth/impl/GithubAuthVendingActorSpec.scala [hunk 11]
```diff
@@ -8,94 +8,81 @@ import cromwell.services.auth.GithubAuthVending.{
...
       )
-    )
 
-    eventually {
-      serviceRegistryActor.send(actor, GithubAuthRequest("invalid_user_token"))
-      serviceRegistryActor.expectMsg(GithubAuthVendingFailure("Exception thrown for testing purposes"))
+      eventually {
+        serviceRegistryActor.send(actor, GithubAuthRequest(TerraToken(userTerraToken)))
+        serviceRegistryActor.expectMsg(expectedResponseMsg)
+      }
```
Comment (cjllanwarne) [inline services/src/test/scala/cromwell/services/auth/impl/GithubAuthVendingActorSpec.scala:85]: You probably want to move the `send` outside the eventually. Otherwise every time the eventually runs, another message is sent

**Review by cjllanwarne — APPROVED** [review]
Looking good, only a few minor comments left

File: services/src/main/scala/cromwell/services/auth/ecm/EcmService.scala (hunk 5 above)
Comment (jgainerdewar) [reply services/src/main/scala/cromwell/services/auth/ecm/EcmService.scala:40]: It's possible that we're hitting this error in cases where we aren't choosing to read the bytes at all, rather than cases where we take too long to read them. We're planning to do [WX-1525](https://broadworkbench.atlassian.net/browse/WX-1525) next week, which should confirm that one way or the other.

File: core/src/main/resources/reference.conf [hunk 12]
```diff
@@ -578,10 +578,7 @@ services {
     config {
       enabled = false
       auth.azure = false
-      # Notes:
-      #   - don't include the 'Bearer' before the token
-      #   - this config value should be removed when support for fetching tokens from ECM has been added to Cromwell
-      access-token = "dummy-token"
+      ecm.base-url = "https://externalcreds.dsde-dev.broadinstitute.org"
```
[the remark this answers is not in the document]
Comment (salonishah11) [reply core/src/main/resources/reference.conf:581]: I see. In this case should there be no default. I can comment this out and add comment describing what service it is looking for?

[the remark this answers is not in the document]
Comment (salonishah11) [reply core/src/main/resources/reference.conf:581]: Something like this?
```
config {
      enabled = false
      auth.azure = false
      # Set this to the service that Cromwell should retrieve Github access token associated with user's token.
      # ecm.base-url = "https://example.org"
    }
```

File: services/src/test/scala/cromwell/services/auth/impl/GithubAuthVendingActorSpec.scala (hunk 10 above)
Comment (salonishah11) [reply services/src/test/scala/cromwell/services/auth/impl/GithubAuthVendingActorSpec.scala:46]: Do you think this is covered by the last test case `return failure message if ECM service returns non-successful response`? It seems that ECM would return a non-successful response if there was no Github token associated with the user and the last test case does simulate `getGithubAccessToken()` returning a Failed future which is what would happen in "no token for user in ECM" case too.

File: services/src/main/scala/cromwell/services/auth/ecm/EcmService.scala (hunk 5 above)
Comment (salonishah11) [reply services/src/main/scala/cromwell/services/auth/ecm/EcmService.scala:40]: In this case, should the whole response be loaded into memory as `Strict` entity using `.toStrict(<timeout>)` method?

File: services/src/main/scala/cromwell/services/auth/impl/GithubAuthVendingActor.scala (hunk 3 above)
Comment (salonishah11) [reply services/src/main/scala/cromwell/services/auth/impl/GithubAuthVendingActor.scala]: Based on [Janet's comment](https://github.com/broadinstitute/cromwell/pull/7392#discussion_r1544673844), I have updated the reference.conf file and commented out `ecm.base-url`. Hence base url is now optional and so I have updated EcmConfig to have Optional baseUrl instead of EcmConfig object being Optional. Since L28 now returns EcmConfig object in all cases I have kept the `apply` method name as is. See [ea97924](https://github.com/broadinstitute/cromwell/pull/7392/commits/ea979242c138e22e28c1ed40f4308dd7bc688546)

File: services/src/main/scala/cromwell/services/auth/impl/GithubAuthVendingActor.scala (hunk 8 above)
Comment (salonishah11) [reply services/src/main/scala/cromwell/services/auth/impl/GithubAuthVendingActor.scala]: done!

File: services/src/main/scala/cromwell/services/auth/impl/GithubAuthVendingActor.scala (hunk 9 above)
Comment (salonishah11) [reply services/src/main/scala/cromwell/services/auth/impl/GithubAuthVendingActor.scala]: done!

File: services/src/test/scala/cromwell/services/auth/impl/GithubAuthVendingActorSpec.scala (hunk 11 above)
Comment (salonishah11) [reply services/src/test/scala/cromwell/services/auth/impl/GithubAuthVendingActorSpec.scala]: done!

File: services/src/main/scala/cromwell/services/auth/ecm/EcmService.scala (hunk 5 above)
Comment (jgainerdewar) [reply services/src/main/scala/cromwell/services/auth/ecm/EcmService.scala:41]: I would just make sure that `responseEntity.dataBytes.runFold` is actually getting executed in all code paths. It looks like you're calling it on both success and failure codes, so I think you're probably okay!

File: core/src/main/resources/reference.conf (hunk 12 above)
[the remark this answers is not in the document]
Comment (jgainerdewar) [reply core/src/main/resources/reference.conf]: Yeah, that looks good!

File: core/src/main/resources/reference.conf [hunk 13]
```diff
@@ -578,10 +578,8 @@ services {
     config {
       enabled = false
       auth.azure = false
-      # Notes:
-      #   - don't include the 'Bearer' before the token
-      #   - this config value should be removed when support for fetching tokens from ECM has been added to Cromwell
-      access-token = "dummy-token"
+      # Set this to the service that Cromwell should retrieve Github access token associated with user's token.
+      # ecm.base-url = "https://externalcreds.dsde-dev.broadinstitute.org/"
```
Comment (jgainerdewar) [inline core/src/main/resources/reference.conf:582]: My previous comment disappeared because of another change to this file (thanks Github) so just calling out that we talked about removing this URL in a now-outdated thread.

Comment (salonishah11) [reply core/src/main/resources/reference.conf]: Oops, I had the change locally but forgot to push it. `ecm.base-url` should now have value `https://example.org`

**Review by jgainerdewar — APPROVED** [review]

File: core/src/main/resources/reference.conf [hunk 14]
```diff
@@ -578,7 +578,8 @@ services {
     config {
       enabled = false
       auth.azure = false
-      ecm.base-url = "https://externalcreds.dsde-dev.broadinstitute.org"
+      # Set this to the service that Cromwell should retrieve Github access token associated with user's token.
+      # ecm.base-url = "https://example.org"
```
Comment (cjllanwarne) [inline core/src/main/resources/reference.conf:582]: Really minor: Can we just leave this blank?

File: services/src/main/scala/cromwell/services/auth/ecm/EcmConfig.scala [hunk 15]
```diff
@@ -3,8 +3,8 @@ package cromwell.services.auth.ecm
 import com.typesafe.config.Config
 import net.ceedubs.ficus.Ficus._
 
-final case class EcmConfig(baseUrl: String)
+final case class EcmConfig(baseUrl: Option[String])
```
Comment (cjllanwarne) [inline services/src/main/scala/cromwell/services/auth/ecm/EcmConfig.scala:6]: TOL is there a reason the `enabled` and `auth.azure` fields are not here?

File: services/src/main/scala/cromwell/services/auth/ecm/EcmConfig.scala [hunk 16]
```diff
@@ -3,8 +3,8 @@ package cromwell.services.auth.ecm
...
 
 object EcmConfig {
-  def apply(config: Config): Option[EcmConfig] = config.as[Option[String]]("ecm.base-url").map(EcmConfig(_))
+  def apply(config: Config): EcmConfig = EcmConfig(config.as[Option[String]]("ecm.base-url"))
```
Comment (cjllanwarne) [inline services/src/main/scala/cromwell/services/auth/ecm/EcmConfig.scala:9]: What if enabled is false?

File: services/src/test/scala/cromwell/services/auth/ecm/EcmConfigSpec.scala [hunk 17]
```diff
@@ -13,10 +13,10 @@ class EcmConfigSpec extends AnyFlatSpec with Matchers {
...
-    val actualEcmConfig = EcmConfig.apply(config)
+    val actualEcmConfig = EcmConfig(config)
 
-    actualEcmConfig shouldBe defined
-    actualEcmConfig.get.baseUrl shouldBe "https://mock-ecm-url.org"
+    actualEcmConfig.baseUrl shouldBe defined
+    actualEcmConfig.baseUrl.get shouldBe "https://mock-ecm-url.org"
   }
 
   it should "return None when ECM base url is absent" in {
```
Comment (cjllanwarne) [inline services/src/test/scala/cromwell/services/auth/ecm/EcmConfigSpec.scala:22]: Should we also have a case for the other config values (to continue the theme... like `enabled`!)

**Review by cjllanwarne — COMMENTED** [review]
I do just have some quick, minor questions about the config layout and "enabled" value

**Review by cjllanwarne — APPROVED** [review]

File: services/src/main/scala/cromwell/services/auth/ecm/EcmConfig.scala (hunk 15 above)
Comment (salonishah11) [reply services/src/main/scala/cromwell/services/auth/ecm/EcmConfig.scala:6]: Discussed via Slack but posting here for posterity: Since these values are more related to enabling `GithubAuthVending` service they are separate from the `ecm` sub-section in the config.

File: services/src/main/scala/cromwell/services/auth/ecm/EcmConfig.scala (hunk 16 above)
Comment (salonishah11) [reply services/src/main/scala/cromwell/services/auth/ecm/EcmConfig.scala:9]: Discussed via Slack but posting here for posterity: Since it is separate from ecm config it is handled separately [here](https://github.com/broadinstitute/cromwell/pull/7392/files#diff-a5fa21faa166c9f13f5872ce92ce2058189460f0c75cef4a3b885d119a50a586R46)

File: services/src/test/scala/cromwell/services/auth/ecm/EcmConfigSpec.scala (hunk 17 above)
Comment (salonishah11) [reply services/src/test/scala/cromwell/services/auth/ecm/EcmConfigSpec.scala:22]: Yes they should be covered [here](https://github.com/broadinstitute/cromwell/pull/7392/files#diff-5af26f8177b51a07147f7e4eb5db049a8a8d3b2f32c58ff7d09d46f87cd397f8R47)

**Comment by salonishah11** [comment]
Not sure why codecov/patch is reporting a low number. I have added unit tests for the modified code as possible but it doesn't seem to be recognizing it 🤔
